c++/src/kj/string.h

```diff
@@ -554,6 +554,107 @@ template <typename T, typename = decltype(toCharSequence(instance<T&>()))>
 inline auto tryToCharSequence(T* value) { return kj::toCharSequence(*value); }
 inline StringPtr tryToCharSequence(...) { return "(can't stringify)"_kj; }
 // SFINAE to stringify a value if and only if it can be stringified.
+
+template <typename T>
+class HasCappedCapacity_ {
+  using yes = char;
+  using no = yes[2];
+
+  template<typename C> static auto test(void*) -> decltype(C::capacity(), yes{});
+  template<typename C> static no& test(...);
+
+public:
+  static constexpr bool value = sizeof(test<T>(nullptr)) == sizeof(yes);
+};
+template <typename T>
+class HasCappedCapacity_<SignalSafeCharSequence<T>>: public HasCappedCapacity_<T> {};
+template <typename T, size_t N>
+class HasCappedCapacity_<T(&)[N]> { static constexpr bool value = true; };
+template <typename T, size_t N>
+class HasCappedCapacity_<const T(&)[N]> { static constexpr bool value = true; };
+
+template <typename T> static inline constexpr bool hasCappedCapacity() {
+  // This returns true if T::capacity() is a method that exists, false otherwise. This is a
+  // helper function for isToCharSequenceResultingInCappedCapacity as I couldn't figure out how to
+  // make the SFINAE magic less verbose.
+  return HasCappedCapacity_<T>::value;
+}
+
+template <typename T>
+struct IsToCharSequenceResultingInCappedCapacity_ {
+  static constexpr bool value = hasCappedCapacity<decltype(tryToCharSequence(instance<T*>()))>();
```

Comment:
Switched to `hasCappedCapacity<decltype(toCharSequence(instance<T>()))>()` using the reasoning you outlaid from above about using `T` instead of trying to work out the reference types.